File: sol/solution/0101-0200/0103/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-spec zigzag_level_order(Root :: #tree_node{} | null) -> [[integer()]].
zigzag_level_order(null) -> [];
zigzag_level_order(Root) ->
  zigzag([[Root]], [], 0).
zigzag([], Acc, _) -> lists:reverse(Acc);
zigzag([[]|Rest], Acc, Dir) -> zigzag(Rest, Acc, Dir);
zigzag(Level, Acc, Dir) ->
  {Vals, NextLevel} = lists:foldl(fun(Node, {Vs, Ns}) ->
    case Node of
      null -> {Vs, Ns};
      #tree_node{val = V, left = L, right = R} ->
        {[V|Vs], Ns ++ [L, R]}
    end
  end, {[], []}, hd(Level)),
  CleanNext = [N || N <- NextLevel, N =/= null],
  ZigzagVals = case Dir rem 2 of
    0 -> lists:reverse(Vals);
    1 -> Vals
  end,
  zigzag([CleanNext | tl(Level)], [ZigzagVals|Acc], Dir+1).

File: sol/solution/0101-0200/0103/Solution.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end
defmodule Solution do
  @spec zigzag_level_order(root :: TreeNode.t | nil) :: [[integer]]
  def zigzag_level_order(nil), do: []
  def zigzag_level_order(root) do
    zigzag_level_order_bfs([root], 0, [])
  end
  defp zigzag_level_order_bfs([], _level, acc), do: Enum.reverse(acc)
  defp zigzag_level_order_bfs(nodes, level, acc) do
    {vals, next_level} =
      Enum.reduce(nodes, {[], []}, fn
        nil, acc -> acc
        node, {vals, next_level} ->
          {vals ++ [node.val], next_level ++ [node.left, node.right]}
      end)
    vals =
      if rem(level, 2) == 1 do
        Enum.reverse(vals)
      else
        vals
      end
    if vals == [] do
      Enum.reverse(acc)
    else
      zigzag_level_order_bfs(Enum.filter(next_level, & &1), level + 1, [vals | acc])
    end
  end
end